8309697: [TESTBUG] Remove "@requires vm.flagless" from jtreg vectorization tests

pfustc opened the pull request:
This patch removes `@require vm.flagless` annotations from HotSpot jtreg tests in `compiler/vectorization/runner`. All jtreg cases in this folder are invoked by test driver `VectorizationTestRunner.java` which checks both correctness and vectorizability (IR) for each test method. We added flagless requirement before because extra flags may mess with compiler control in the test driver for correctness check. But `flagless` has a side effect that it makes tests with extra flags skipped. So we propose to get rid of it now.

To adapt the removal of `@require vm.flagless`, a few checks are added in the test driver to skip the correctness check if extra flags make the compiler control not work. This patch also moves previously hard-coded flag `-XX:-OptimizeFill` in the test driver to conditions in IR checks.

Tested various of compiler control related VM flags on x86 and AArch64.

<!-- Anything below this marker will be automatically updated, please do not edit manually! -->
---------
### Progress
- [x] Change must be properly reviewed (1 review required, with at least 1 [Reviewer](https://openjdk.org/bylaws#reviewer))
- [x] Change must not contain extraneous whitespace
- [x] Commit message must refer to an issue

### Issue
 * [JDK-8309697](https://bugs.openjdk.org/browse/JDK-8309697): [TESTBUG] Remove "@<!---->requires vm.flagless" from jtreg vectorization tests (**Enhancement** - P4)


### Reviewers
 * [Vladimir Kozlov](https://openjdk.org/census#kvn) (@vnkozlov - **Reviewer**)
 * [Tobias Hartmann](https://openjdk.org/census#thartmann) (@TobiHartmann - **Reviewer**)
 * [Emanuel Peter](https://openjdk.org/census#epeter) (@eme64 - **Reviewer**)
 * [Christian Hagedorn](https://openjdk.org/census#chagedorn) (@chhagedorn - **Reviewer**)


### Reviewing
<details><summary>Using <code>git</code></summary>

Checkout this PR locally: \
`$ git fetch https://git.openjdk.org/jdk.git pull/15011/head:pull/15011` \
`$ git checkout pull/15011`

Update a local copy of the PR: \
`$ git checkout pull/15011` \
`$ git pull https://git.openjdk.org/jdk.git pull/15011/head`

</details>
<details><summary>Using Skara CLI tools</summary>

Checkout this PR locally: \
`$ git pr checkout 15011`

View PR using the GUI difftool: \
`$ git pr show -t 15011`

</details>
<details><summary>Using diff file</summary>

Download this PR as a diff file: \
<a href="https://git.openjdk.org/jdk/pull/15011.diff">https://git.openjdk.org/jdk/pull/15011.diff</a>

</details>


### Webrev
[Link to Webrev Comment](https://git.openjdk.org/jdk/pull/15011#issuecomment-1649388940)

pfustc commented:
Hi @eme64 @vnkozlov, This removes flagless from jtreg vectorization tests as you previously suggested. Could you help take a look?

vnkozlov commented on test/hotspot/jtreg/compiler/vectorization/runner/VectorizationTestRunner.java:
```diff
@@ -131,7 +138,10 @@ private void runTestOnMethod(Method method) throws InterruptedException {
         WB.enqueueMethodForCompilation(method, COMP_LEVEL_C2);
         while (WB.getMethodCompilationLevel(method) != COMP_LEVEL_C2) {
             if (System.currentTimeMillis() - enqueueTime > COMP_THRES_SECONDS * 1000) {
-                fail("Method is not compiled after " + COMP_THRES_SECONDS + "s.");
+                // C2 compilation may timeout with extra compiler control options.
+                // We skip the correctness check in this case.
+                System.out.println("WARNING: Correctness check is skipped because C2 compilation times out");
+                return;
```
Why not run with `BackgroundCompilation` OFF? The code will return from `WHITE_BOX.enqueueMethodForCompilation()` only after finishing compiling or compiler skip.
Compilation speed is very different on different platform. Using time for check is not reliable.

pfustc replied:
> Why not run with BackgroundCompilation OFF?

I just explained the reason in my another reply above. Using time for check is to avoid jtreg timeout with extra VM options like `-XX:CompileThreshold=100 -XX:-TieredCompilation`. But I think this can also be avoided if we get reference result from C1 execution. What do you think of the new approach?

TobiHartmann commented:
`LoopArrayIndexComputeTest.java` fails on x64 with `-XX:UseAVX=0 -XX:UseSSE=3` and with `-XX:UseAVX=0 -XX:UseSSE=2`:

```
Failed IR Rules (1) of Methods (1)
----------------------------------
1) Method "public byte[] compiler.vectorization.runner.LoopArrayIndexComputeTest.byteArrayWithDependencePos()" - [Failed IR rules: 1]:
   * @IR rule 1: "@compiler.lib.ir_framework.IR(applyIfCPUFeatureAnd={}, phase={DEFAULT}, applyIfCPUFeatureOr={"asimd", "true", "sse2", "true"}, applyIf={"AlignVector", "false"}, applyIfCPUFeature={}, counts={"_#STORE_VECTOR#_", ">0"}, failOn={}, applyIfAnd={}, applyIfOr={}, applyIfNot={})"
     > Phase "PrintIdeal":
       - counts: Graph contains wrong number of nodes:
         * Constraint 1: "(\\d+(\\s){2}(StoreVector.*)+(\\s){2}===.*)"
           - Failed comparison: [found] 0 > 0 [given]
           - No nodes matched!
```

pfustc commented on test/hotspot/jtreg/compiler/vectorization/runner/VectorizationTestRunner.java:
```diff
@@ -54,31 +54,38 @@ public class VectorizationTestRunner {
-                    runTestOnMethod(method);
+        // Some VM options like "-Xcomp" may mess with the compiler control for the
+        // correctness check. We disable the check in these cases.
+        boolean use_intp = WB.getBooleanVMFlag("UseInterpreter");
+        boolean use_comp = WB.getBooleanVMFlag("UseCompiler");
+        boolean bg_comp = WB.getBooleanVMFlag("BackgroundCompilation");
+        if (use_intp && use_comp && bg_comp) {
```
For `-Xbatch` we need check `BackgroundCompilation`. As we lock the compilation before running the test method in the interpreter to get reference result, the VM will hang if both background compilation is disabled and the compilation is locked by WhiteBox.

I just experimented an alternative approach that running reference result with C1 execution instead of interpretation. In this way we can run with background compilation off. What do you think of this new approach?

pfustc commented:
Hi @TobiHartmann ,

> `LoopArrayIndexComputeTest.java` fails on x64 with `-XX:UseAVX=0 -XX:UseSSE=3` and with `-XX:UseAVX=0 -XX:UseSSE=2`:

I just tried this on multiple x86 machines we have but didn't reproduce the failure. Could you share more info (cpu features, etc.) of your test machine so I can find why the test method is not vectorized.

vnkozlov commented:
> Hi @TobiHartmann ,
> 
> > `LoopArrayIndexComputeTest.java` fails on x64 with `-XX:UseAVX=0 -XX:UseSSE=3` and with `-XX:UseAVX=0 -XX:UseSSE=2`:
> 
> I just tried this on multiple x86 machines we have but didn't reproduce the failure. Could you share more info (cpu features, etc.) of your test machine so I can find why the test method is not vectorized.

The test has Byte multiply operation but it checks only presence of SSE2:
``
@IR(applyIfCPUFeatureOr = {"asimd", "true", "sse2", "true"},
...
res[i] *= bytes[i + 3];
``
But we don't support it with SSE < 4 - see `Matcher::match_rule_supported()`.

`vm.flagless` prevented running it before with such flags combination.

I am not sure how your testing passed.

pfustc commented:
Hi @vnkozlov ,

Thanks for your reply. But it still has problems.

> About your change to allow -Xbatch. Let me clarify, if you exclude -Xcomp mode (which I agree with) by checking UseInterpreter flag for true, then a method could be always executed in Interpeter to get reference result (even with -XX:CompileThreshold=100) by calling method once first (we do that in other tests).

> You don't need to call WB.lockCompilation() if you exclude -Xcomp mode. There will be no compilation requests for called method when you call the method first time because compilation threshold will not be reached - it is guarantee that method will be executed in Interpreter. And you have the assert to verify that.

These tests are a bit different because we test loops. If the loop iteration count reaches some threshold, the loop will be *OSR compiled* even test method is called only once. I just did an experiment according to your suggestion. After removing `WB.lockCompilation()` and updating loop iteration count to 100,000, I got assertion failure that tells me the test method is NOT running in interpreter.

```
STDERR:
java.lang.AssertionError
	at compiler.vectorization.runner.VectorizationTestRunner.runTestOnMethod(VectorizationTestRunner.java:131)
	at compiler.vectorization.runner.VectorizationTestRunner.run(VectorizationTestRunner.java:73)
	at compiler.vectorization.runner.VectorizationTestRunner.main(VectorizationTestRunner.java:215)
	at java.base/jdk.internal.reflect.DirectMethodHandleAccessor.invoke(DirectMethodHandleAccessor.java:103)
	at java.base/java.lang.reflect.Method.invoke(Method.java:580)
	at com.sun.javatest.regtest.agent.MainWrapper$MainTask.run(MainWrapper.java:138)
	at java.base/java.lang.Thread.run(Thread.java:1570)
```

A solution to this may be adding one more check of `CICompileOSR` is OFF if we still want to use interpreted execution for the reference result.

Now the question is, which verification approach do you think is better? "C2 vs. interpreted" or "C2 vs. C1"?

vnkozlov commented:
> A solution to this may be adding one more check of `CICompileOSR` is OFF if we still want to use interpreted execution for the reference result.

I would suggest to use `WB.setBooleanVMFlag("CICompileOSR", false);`. But it is debug flag which can be set only in debug VM.  There are may be other product flags you can temporary set to avoid compilation without locking.

> 
> Now the question is, which verification approach do you think is better? "C2 vs. interpreted" or "C2 vs. C1"?

We usually use Interpreter as gold standard.

pfustc commented:
Hi @vnkozlov ,

Thanks for your comments. I have reverted the patch to my 1st commit and re-addressed your comments.

> I would suggest to use WB.setBooleanVMFlag("CICompileOSR", false);. But it is debug flag which can be set only in debug VM. There are may be other product flags you can temporary set to avoid compilation without locking.

In my new commit, I choose to set and restore `UseCompiler` before and after the interpreter run. I have re-tested various of compiler control options and no jtreg timeout is seen now. Please let me know if this looks good to you.

vnkozlov reviewed (approved):
This looks good.

eme64 commented:
@vnkozlov @TobiHartmann we should re-run testing from our side.

@pfustc Why do you only test correctness (compare results) in some conditions? Is there not a risk that we miss doing it in some cases we should do it, just because we get the conditions slightly wrong?

Just FYI: we should integrate this whole correctness of results testing into the IR framework. I filed [JDK-8310533](https://bugs.openjdk.org/browse/JDK-8310533). That would make it easier to use for new tests. It could also be used for any test, not just the ones located in `test/hotspot/jtreg/compiler/vectorization`.

pfustc commented:
Hi @eme64 ,

Thanks for looking at this.

> @pfustc Why do you only test correctness (compare results) in some conditions? Is there not a risk that we miss doing it in some cases we should do it, just because we get the conditions slightly wrong?

Yes, you are right! These conditions are added before to avoid jtreg hanging when compilation is locked. But now I can remove them because the lock is removed. In my latest commit, I have removed the conditions and some useless imports.

> Just FYI: we should integrate this whole correctness of results testing into the IR framework. I filed [JDK-8310533](https://bugs.openjdk.org/browse/JDK-8310533). That would make it easier to use for new tests. It could also be used for any test, not just the ones located in test/hotspot/jtreg/compiler/vectorization.

I have noticed this JBS before. The reason I didn't added correctness check into the IR framework is that I implemented this kind of check before the IR framework exists. (We have used it internally for a few years.) But anyway, it is a good proposal and I'm willing to help if needed.

eme64 commented:
@pfustc This looks good to me, thanks for making these changes. It will really increase the coverage.
Maybe @vnkozlov should quickly look at it again if he still agrees.
@TobiHartmann is running the testing again, just in case the dropped conditions change something.
I will give you my approval after those tests are passing.

vnkozlov reviewed (approved):
I am fine with update.

TobiHartmann commented:
All tests passed (@eme64).

eme64 reviewed (approved):
Thanks @TobiHartmann .

-> approved

eme64 commented:
@pfustc Since you are working on IR tests here, it would be good if you first merge the changes from [JDK-8310308](https://bugs.openjdk.org/browse/JDK-8310308), and test the IR rules again.

pfustc commented:
Thanks @vnkozlov @TobiHartmann @eme64 for your review. Patch is currently merged with latest master and re-tested. I will integrate this if there's no further comment.

chhagedorn commented:
I will also give this another spin in our testing.

chhagedorn reviewed (approved):
Testing looked good!

pfustc commented:
/integrate
